horace_core/sqw/@sqw/parse_pixel_indices.m

```diff
@@ -1,212 +1,206 @@
+        % Outputs are cell arrays of numeric arrays
+        [irun, idet, ien] = arrayfun(@parse_pixel_indices_private, win, ...
+            'uniformOutput', false);
     end
 end
 
-% Fill output arrays if required
-if nout>0
-    irun = cell(size(win));
-    idet = cell(size(win));
-    ien  = cell(size(win));
-
-    for i = 1:nw
-        indx_tmp = indx_internal{iw_internal(i)};
-        if size(indx_tmp,2)==1
-            if iscell(win)
-                pix = win{i}.pix;
-                runid_map = win{i}.runid_map;
-            else
-                pix = win(i).pix;
-                runid_map = win(i).runid_map;
-            end
+end
 
-            run_idx= pix.run_idx(indx_tmp)';
-            irun{i} = arrayfun(@(x)runid_map(x),run_idx);
-            idet{i} = pix.detector_idx(indx_tmp)';
-            ien{i}  = pix.energy_idx(indx_tmp)';
-        else
-            irun{i} = indx_tmp(:,1);
-            %? Does this index need to be transformed from run_idx to
-            %index ?
-            idet{i} = indx_tmp(:,2);
-            ien{i}  = indx_tmp(:,3);
+%-------------------------------------------------------------------------------
+function [irun, idet, ien] = parse_pixel_indices_private (win, ipix)
+% Get the experiment indices (sometimes known as 'header index'), detector
+% indices and energy bin indices for an sqw object.
+% Private function that assumes all validity checks for the input arguments have
+% already been done.
+%
+% Input:
+% ------
+%   win         sqw object
+%   ipix        Pixel indices array, or cell array with one element that is an
+%               array of pixel indices
+%
+% Output:
+% -------
+%   irun -|
+%   idet  |-    Indices corresponding to the selected pixels (column vectors)
+%   ien  -|
+
+pix = win.pix;
+experiment = win.experiment_info;
+
+if pix.num_pixels > 0
+    % At least one pixel, so need to do some work
+    if exist('ipix', 'var')
+        % Get irun, idet, ien for pixels indicated by ipix with same shape as ipix
+        if iscell(ipix)
+            ipix = ipix{1};     % get the indices array inside ipix
         end
+        run_idx = reshape(pix.run_idx(ipix), size(ipix));
+        irun = experiment.get_experiment_idx(run_idx);
+        idet = reshape(pix.detector_idx(ipix), size(ipix));
+        ien = reshape(pix.energy_idx(ipix), size(ipix));
+    else
+        % Get irun, idet, ien for all pixels in the sqw object
+        run_idx = pix.run_idx(:);   % make column
+        irun = experiment.get_experiment_idx(run_idx); % irun same shape as run_idx i.e. column
+        idet = pix.detector_idx(:); % make column
+        ien = pix.energy_idx(:);    % make column
     end
-
-    if nw == 1
-        irun = irun{1};
-        idet = idet{1};
-        ien = ien{1};
-    end
+else
+    % Catch the case of no pixels for simplicity
+    irun = zeros(0,1);  % convention is column vector
+    idet = zeros(0,1);
+    ien = zeros(0,1);
 end
```

Comment:
Depending on your thoughts on single returns from functions (single point of exit at the bottom) I would probably move this up as an early `return` to flatten this function somewhat.
```
if pix.num_pixels == 0  % Catch the case of no pixels for simplicity
    irun = zeros(0,1);  % convention is column vector
    idet = zeros(0,1);
    ien = zeros(0,1);
    return;
end
```
Or flatten the conditions as:
```
if pix.num_pixels == 0  % Catch the case of no pixels for simplicity
   ...
elseif exist('ipix', 'var')
   ...
else
   ...
end
```

Reply:
Usually I prefer to have a single point of exit at the end, but in this case I think that it improves the readability considerably to catch the trivial case first and returning, as you suggest in option 1. I didn't flatten the if statement as in your second option, as it places two very different tests (no pixels, and the presence of an optional input argument) into one if...end block. I think it better to separate the two tests.